File: aac_library/src/main/java/com/example/aac_library/base/view/BaseActivity.java
package com.example.aac_library.base.view;

import android.app.ProgressDialog;
import android.os.Bundle;
import android.widget.Toast;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModel;
import com.example.aac_library.base.interf.IView;
import com.example.aac_library.base.interf.IViewModel;
import com.example.aac_library.base.event.BaseActionEvent;

import java.util.ArrayList;
import java.util.List;

public abstract class BaseActivity extends AppCompatActivity implements IView {
    private ProgressDialog progressDialog;
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(initLayoutResId());
        initDataBinding();
        initViewModelEvent();
        initData();
    }



    private void initViewModelEvent(){
        List<ViewModel> viewModels  = initViewModels();
        if (viewModels != null && viewModels.size()>0) {
            observeEvent(viewModels);
        }else {
            ViewModel viewModel = initViewModel();
            if (viewModel != null) {
                List<ViewModel> modelList = new ArrayList<>();
                modelList.add(viewModel);
                observeEvent(modelList);
            }
        }

    }

    private void observeEvent(List<ViewModel> viewModels){

        for (ViewModel viewModel : viewModels) {
            if (viewModel instanceof IViewModel){
                IViewModel iViewModelAction = (IViewModel) viewModel;
                iViewModelAction.getActionLiveData().observe(this, baseActionEvent -> {
                    switch (baseActionEvent.getAction()){
                        case BaseActionEvent.SHOW_LOADING_DIALOG:
                            startLoading(baseActionEvent.getMessage());
                            break;
                        case BaseActionEvent.DISMISS_LOADING_DIALOG:
                            dismissLoading();
                            break;
                        case BaseActionEvent.SHOW_TOAST:
                            showToast(baseActionEvent.getMessage());
                            break;
                        case BaseActionEvent.FINISH:
                            break;
                        case BaseActionEvent.FINISH_WITH_RESULT_OK:
                            finishWithResultOk();
                            break;
                    }
                });
            }
        }
    }

    protected abstract ViewModel initViewModel();

    protected abstract int initLayoutResId();

    protected abstract void initData();

    protected abstract void initDataBinding();
    private BaseActivity getContext(){
        return this;
    }
    protected List<ViewModel>  initViewModels(){
        return null;
    }


    @Override
    public void startLoading() {
        startLoading("");
    }

    @Override
    public void startLoading(String msg) {

        if (progressDialog != null) {
            progressDialog.setTitle(msg);
            progressDialog.show();
        }else {
            progressDialog = new ProgressDialog(this);
            progressDialog.setCancelable(false);
            progressDialog.setCanceledOnTouchOutside(false);
        }
    }

    @Override
    public void dismissLoading() {

        if (progressDialog != null && progressDialog.isShowing()) {
            progressDialog.dismiss();
        }

    }

    @Override
    public void showToast(String msg) {

        Toast.makeText(this,msg,Toast.LENGTH_SHORT).show();
    }


    @Override
    public void finishWithResultOk() {

    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        dismissLoading();
    }
}

File: aac_library/src/main/java/com/example/aac_library/base/view/BaseActivity.kt
package com.example.aac_library.base.view

import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import com.example.aac_library.base.BaseViewModel
import com.example.aac_library.base.event.BaseEvent
import com.example.aac_library.base.interf.IBaseView
import com.sunchen.netbus.NetStatusBus
import java.util.ArrayList

abstract class BaseActivity : AppCompatActivity(), IBaseView {
    private val context: BaseActivity
        get() = this

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(initLayoutResId())
        initDataBinding()
        initViewModelEvent()
        initView()
        initData()
    }


    protected abstract fun initLayoutResId(): Int
    protected abstract fun initView()
    protected abstract fun initData()
    protected abstract fun initDataBinding()

    protected abstract fun initViewModel(): BaseViewModel?
    /**
     * 暂时没有多个view model的情况
     */
    private fun initViewModels():List<BaseViewModel>?{
        return null
    }
    override fun startLoading() {
        startLoading("")
    }

    override fun startLoading(msg: String) {
    }

    override fun dismissLoading() {
    }

    override fun showToast(msg: String) {
        Toast.makeText(this,msg,Toast.LENGTH_SHORT).show()
    }

    private fun initViewModelEvent() {
        val viewModels = initViewModels()
        if (viewModels != null && viewModels.isNotEmpty()) {
            observeEvent(viewModels)
        } else {
            val viewModel = initViewModel()
            if (viewModel != null) {
                val modelList = ArrayList<BaseViewModel>()
                modelList.add(viewModel)
                observeEvent(modelList)
            }
        }

    }

    private fun observeEvent(viewModels: List<BaseViewModel>) {

        for (viewModel in viewModels) {

            viewModel.actionLiveData.observe(this, Observer{ baseActionEvent ->
                baseActionEvent?.let {
                    when (it.action) {
                        BaseEvent.SHOW_LOADING_DIALOG -> startLoading(it.message!!)
                        BaseEvent.DISMISS_LOADING_DIALOG -> dismissLoading()
                        BaseEvent.SHOW_TOAST -> showToast(it.message!!)
                        BaseEvent.FINISH -> finish()
                        BaseEvent.FINISH_WITH_RESULT_OK -> finish()
                    }
                }

            })
        }
    }

    fun bindService(serviceClass: Class<*>,connection: ServiceConnection) {
        bindService(Intent(this, serviceClass), connection, Context.BIND_AUTO_CREATE)
    }
    fun unBindService(connection: ServiceConnection) {
        unbindService(connection)
    }
    fun reBindService(serviceClass: Class<*>,connection: ServiceConnection) {
        unbindService(connection)
        bindService(Intent(this, serviceClass), connection, Context.BIND_AUTO_CREATE)
    }

    public override fun onStart() {
        super.onStart()
        NetStatusBus.getInstance().register(this)
    }

    public override fun onStop() {
        super.onStop()
        NetStatusBus.getInstance().unregister(this)
    }

    override fun onDestroy() {
        super.onDestroy()
        dismissLoading()
    }
}
